### ISSUE TITLE
Using `volatile` for thread sync is not safe

### ISSUE THREAD

**Issue by jktjkt** [CESNET/Netopeer2#405]
I added netopeer2-server to our CI and immediately noticed that about one in twenty test suite runs gets stuck. Investigating with a debugger I found out that the stuck test instance was spinning in a loop, waiting for `initialized` to be set.

[`volatile` is not a tool for thread safety](https://stackoverflow.com/questions/2484980/why-is-volatile-not-considered-useful-in-multithreaded-c-or-c-programming), its purpose is handling of special memory regions such as memory-mapped IO. It is not safe to have a `volatile int` set from one thread and checked from the other one. This pattern is present in the test suite and in the server's main control loop.

I tried to [convert this to semaphores](https://github.com/jktjkt/Netopeer2/commit/5e490aa96fc62c844609dc68706fcd37ebbd6339), but it seems that this is not enough. Some of these tests still get stuck, occasionally:

```
(gdb) thread apply all bt

Thread 1 (Thread 0x7f401d86f940 (LWP 15292)):
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x55aff408e0c0 <wait_for_init>) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x55aff408e0c0 <wait_for_init>, abstime=0x0) at sem_waitcommon.c:111
#2  0x00007f401c2baaa8 in __new_sem_wait_slow (sem=0x55aff408e0c0 <wait_for_init>, abstime=0x0) at sem_waitcommon.c:181
#3  0x000055aff3da6e3a in np_start (state=0x610000000050) at /home/jkt/work/cesnet/gerrit/CzechLight/cla-sysrepo/submodules/dependencies/Netopeer2/server/tests/test_delete_config.c:598
#4  0x00007f401c4cd4e3 in cmocka_run_one_test_or_fixture (function_name=0x55aff3dc8840 "test_startstop", test_func=test_func@entry=0x0, 
    setup_func=setup_func@entry=0x55aff3da6d91 <np_start>, teardown_func=teardown_func@entry=0x0, state=<optimized out>, state@entry=0x610000000050, heap_check_point=<optimized out>)
    at /var/tmp/portage/dev-util/cmocka-1.0.1/work/cmocka-1.0.1/src/cmocka.c:2310
#5  0x00007f401c4cdc28 in cmocka_run_one_tests (test_state=0x610000000040) at /var/tmp/portage/dev-util/cmocka-1.0.1/work/cmocka-1.0.1/src/cmocka.c:2395
#6  _cmocka_run_group_tests (group_name=0x55aff3dc89e0 "tests", tests=<optimized out>, num_tests=4, group_setup=<optimized out>, group_teardown=0x0)
    at /var/tmp/portage/dev-util/cmocka-1.0.1/work/cmocka-1.0.1/src/cmocka.c:2518
#7  0x000055aff3da773a in main () at /home/jkt/work/cesnet/gerrit/CzechLight/cla-sysrepo/submodules/dependencies/Netopeer2/server/tests/test_delete_config.c:730
```

Can you please give it a try?

**Comment by michalvasko** [comment]
Hi,
so the same old story once again, but maybe you will provide some new insight.

The two properties of `volatile` taken from the question you referenced:
> the read/write actually happens (that the compiler won't just store the value in a register instead and defer updating main memory until much later)

> guarantees that no reordering occurs between different volatile reads/writes

We can use `test_close_session.c` as an example. Going sequentially, `initialized` is first set to `0`, then a thread is created that tests `initialized` and if not set, sets it to `1`. Once it does, the calling function `np_start()` returns.

Now, the `volatile` purpose here is to make the `initialized` value change visible for `np_start()` (the first property mentioned). If that happens, the infinite loop **must** exit. As for the second property, I do not think it is even relevant in this case because it is not used to access some other data or anything else, it is just a flag.

Do you think I have said something untrue? If not, can you please explain how it is even theoretically possible for the loop to get stuck?

Regards,
Michal

**Comment by jktjkt** [comment]
You just described the bug -- thread 2 sets a variable to `1` while thread 1 busy-waits on the same variable, waiting for it to become `1`. That satisfies the definition of an unsynchronized, concurrent access from two threads, and that is a bug. Keep in mind that both the compiler and optimizer, the CPU and the memory controller are all free to reorder other memory reads and writes "around" the load/store to the `volatile` memory. It is not "just a flag", it is a flag which attempts to synchronize two threads which both perform side-effects.

**Comment by jktjkt** [comment]
If you want access to the CI logs which show the stuck test suite, just ask.

**Comment by michalvasko** [comment]
Hi,
could you try the latest `devel` (I forgot to reference this issue)? I have simply used atomic variables instead.

Regards,
Michal

**Comment by jktjkt** [comment]
I do not think that `memory_order_relaxed` is a correct choice here. The docs say [that this is not a synchronization primitive](https://en.cppreference.com/w/c/atomic/memory_order#Relaxed_ordering):

> Atomic operations tagged `memory_order_relaxed` are not synchronization operations; they do not impose an order among concurrent memory accesses. They only guarantee atomicity and modification order consistency. 

And it seems that clang's `-fsanitize=thread` agrees with me, for example for the `test_close_session`:
```
==================
WARNING: ThreadSanitizer: data race (pid=10602)
  Read of size 4 at 0x000000f61034 by main thread:
    #0 test_close_session /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:267 (test_close_session+0x4f57ed)
    #1 _test_realloc ??:? (libcmocka.so.0+0x5426)
    #2 __libc_start_main ??:? (libc.so.6+0x24412)

  Previous write of size 4 at 0x000000f61034 by thread T1 (mutexes: read M22):
    #0 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:157 (test_close_session+0x4f4cd7)
    #1 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f394b)
    #2 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #3 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  As if synchronized via sleep:
    #0 usleep ??:? (test_close_session+0x43e3b9)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:239 (test_close_session+0x4f5925)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

  Location is global '<null>' at 0x000000000000 (test_close_session+0x000000f61034)

  Mutex M22 (0x000000f610f0) created at:
    #0 pthread_rwlock_init ??:? (test_close_session+0x442f77)
    #1 np2srv_init_schemas /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:964 (test_close_session+0x4f6b52)
    #2 server_init /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1258 (test_close_session+0x4f2ec8)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1739 (test_close_session+0x4f2203)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  Thread T1 (tid=10615, running) created by main thread at:
    #0 pthread_create ??:? (test_close_session+0x42df5a)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:236 (test_close_session+0x4f58ce)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

SUMMARY: ThreadSanitizer: data race /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:267 in test_close_session
==================
==================
WARNING: ThreadSanitizer: data race (pid=10602)
  Read of size 8 at 0x7ba000000070 by main thread:
    #0 write ??:? (test_close_session+0x451f6d)
    #1 test_write /home/ci/build/Netopeer2/server/./tests/config.h:342 (test_close_session+0x4f9068)
    #2 test_close_session /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:267 (test_close_session+0x4f5802)
    #3 _test_realloc ??:? (libcmocka.so.0+0x5426)
    #4 __libc_start_main ??:? (libc.so.6+0x24412)

  Previous write of size 8 at 0x7ba000000070 by thread T1 (mutexes: read M22):
    #0 pipe ??:? (test_close_session+0x448e02)
    #1 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:149 (test_close_session+0x4f4bad)
    #2 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f394b)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  As if synchronized via sleep:
    #0 usleep ??:? (test_close_session+0x43e3b9)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:239 (test_close_session+0x4f5925)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

  Location is file descriptor 7 created by thread T1 at:
    #0 pipe ??:? (test_close_session+0x448e02)
    #1 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:149 (test_close_session+0x4f4bad)
    #2 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f394b)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  Mutex M22 (0x000000f610f0) created at:
    #0 pthread_rwlock_init ??:? (test_close_session+0x442f77)
    #1 np2srv_init_schemas /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:964 (test_close_session+0x4f6b52)
    #2 server_init /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1258 (test_close_session+0x4f2ec8)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1739 (test_close_session+0x4f2203)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  Thread T1 (tid=10615, running) created by main thread at:
    #0 pthread_create ??:? (test_close_session+0x42df5a)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:236 (test_close_session+0x4f58ce)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

SUMMARY: ThreadSanitizer: data race ??:? in __interceptor_write
==================
==================
WARNING: ThreadSanitizer: data race (pid=10602)
  Read of size 4 at 0x000000f61030 by main thread:
    #0 test_close_session /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:268 (test_close_session+0x4f5814)
    #1 _test_realloc ??:? (libcmocka.so.0+0x5426)
    #2 __libc_start_main ??:? (libc.so.6+0x24412)

  Previous write of size 4 at 0x000000f61030 by thread T1 (mutexes: read M22):
    #0 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:156 (test_close_session+0x4f4c9b)
    #1 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f394b)
    #2 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #3 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  As if synchronized via sleep:
    #0 usleep ??:? (test_close_session+0x43e3b9)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:239 (test_close_session+0x4f5925)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

  Location is global '<null>' at 0x000000000000 (test_close_session+0x000000f61030)

  Mutex M22 (0x000000f610f0) created at:
    #0 pthread_rwlock_init ??:? (test_close_session+0x442f77)
    #1 np2srv_init_schemas /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:964 (test_close_session+0x4f6b52)
    #2 server_init /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1258 (test_close_session+0x4f2ec8)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1739 (test_close_session+0x4f2203)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  Thread T1 (tid=10615, running) created by main thread at:
    #0 pthread_create ??:? (test_close_session+0x42df5a)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:236 (test_close_session+0x4f58ce)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

SUMMARY: ThreadSanitizer: data race /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:268 in test_close_session
==================
==================
WARNING: ThreadSanitizer: data race (pid=10602)
  Read of size 8 at 0x7ba000000040 by main thread:
    #0 read ??:? (test_close_session+0x4567ad)
    #1 test_read /home/ci/build/Netopeer2/server/./tests/config.h:374 (test_close_session+0x4f92fa)
    #2 test_close_session /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:268 (test_close_session+0x4f5829)
    #3 _test_realloc ??:? (libcmocka.so.0+0x5426)
    #4 __libc_start_main ??:? (libc.so.6+0x24412)

  Previous write of size 8 at 0x7ba000000040 by thread T1 (mutexes: read M22):
    #0 pipe ??:? (test_close_session+0x448e02)
    #1 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:148 (test_close_session+0x4f4b97)
    #2 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f394b)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  As if synchronized via sleep:
    #0 usleep ??:? (test_close_session+0x43e3b9)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:239 (test_close_session+0x4f5925)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

  Location is file descriptor 4 created by thread T1 at:
    #0 pipe ??:? (test_close_session+0x448e02)
    #1 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:148 (test_close_session+0x4f4b97)
    #2 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f394b)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  Mutex M22 (0x000000f610f0) created at:
    #0 pthread_rwlock_init ??:? (test_close_session+0x442f77)
    #1 np2srv_init_schemas /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:964 (test_close_session+0x4f6b52)
    #2 server_init /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1258 (test_close_session+0x4f2ec8)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1739 (test_close_session+0x4f2203)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f99f1)

  Thread T1 (tid=10615, running) created by main thread at:
    #0 pthread_create ??:? (test_close_session+0x42df5a)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:236 (test_close_session+0x4f58ce)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

SUMMARY: ThreadSanitizer: data race ??:? in read
==================
ThreadSanitizer: reported 4 warnings
```

I will give the default sequential memory order a try.

**Comment by michalvasko** [comment]
Hi,
you are right, for some reason I completely ignored the fact that the flag **is** used for synchronizing the pipes. Based on my understanding the *acquire-release* ordering should be sufficient because it basically provides a memory fence between the threads synchronized on the particular atomic variable this way. However, the `control` variable in the server itself should be fine with relaxed memory ordering (I have even read somewhere that signaling other threads to quit with a flag is an example use-case of this ordering).

When running the tests after the last changes with `helgrind` I also got some warnings about data races. I believe they are, in short, the result of busy waiting and should cause no harm. What do you think?

Regards,
Michal

**Comment by jktjkt** [comment]
Here's a [list of all warnings currently generated by TSAN](https://ci-logs.gerrit.cesnet.cz/t/public/logs//60/1560/10/check/f29-clang-tsan/6e0c3c6/) (IPv6 required) for building libyang, sysrepo, libnetconf2, netopeer2-server, including their test suite. The build runs as an unprivileged user and sysrepo is built twice, once for repo-dir-outside-build and once in the usual dev mode with repo dir inside the build dir -- see [the build script](https://gerrit.cesnet.cz/plugins/gitiles/CzechLight/dependencies/+/master/ci/build.sh) in case that's relevant.

- I'm ignoring the lock order inversion for now because it's a ton of logs and it's deep within code that I am not familiar with. Generally, if two threads take locks A and B and sometimes B gets locked after A and at other times it's the other way round, there's a real potential for deadlock.
- There's a [race around `lyp_check_date` -> `tzset_internal`](https://ci-logs.gerrit.cesnet.cz/t/public/logs//60/1560/10/check/f29-clang-tsan/6e0c3c6/tsan.log.7804) which is caused by a call to `mktime`. I think that this one might be harmless *if* we expect breakage when someone sets a timezone at just the wrong moment. I can live with that one, but fixing it should be reasonably easy -- it's "just" a matter of writing a mktime-like funciton which doesn't work on local time, but in UTC.
- I have not looked into libnetconf2's `test_fd_comm.c` too carefully ([log](https://ci-logs.gerrit.cesnet.cz/t/public/logs//60/1560/10/check/f29-clang-tsan/6e0c3c6/tsan.log.9915)). I remember seeing two threads manipulating a set of pipe FDs. This should be fixed, eventually, because it is the same class of bug as the `int initialized` in Netopeer2-server's test suite.
- this [`lys_node_free`](https://ci-logs.gerrit.cesnet.cz/t/public/logs//60/1560/10/check/f29-clang-tsan/6e0c3c6/tsan.log.10604) data race looks suspicious and I do not remember seeing it yesterday, so it would be nice if you took a look at it, please
- there's definitely [a bug in libnetconf2's counting of active sessions](https://ci-logs.gerrit.cesnet.cz/t/public/logs//60/1560/10/check/f29-clang-tsan/6e0c3c6/tsan.log.10608) because it is using a plain integer

My preference is for you to take a look at these because I'm quite removed from sysrepo internals and C (as opposed to C++) programming, so I believe that you're better qualified to fix these, and in fewer time. I'll be happy to test any patches or pull requests, it's just a matter of waiting 20 minutes for the build job to finish, and I have tooling in place for this. I could probably fix both of the libnetconf2 issues if you wanted me to, though.

For me, this is not an exercise in blindly fixing warnings. We would like to, in the end, make these TSAN warnings fatal because it makes it easier to just stop the build of our internal software when we hit a new warning, and that requires a reasonably clear execution of the whole dependency stack.

**Comment by michalvasko** [comment]
Hi Jan,
> For me, this is not an exercise in blindly fixing warnings.

You are probably referring to the second paragraph of my last post. This is an example of output I meant
```
==19820== 
==19820== Possible data race during write of size 1 at 0x42CE88 by thread #2
==19820== Locks held: none
==19820==    at 0x40EB81: __wrap_nc_accept (test_close_session.c:159)
==19820==    by 0x40DBE9: worker_thread (main.c:1400)
==19820==    by 0x40E880: server_main (main.c:1753)
==19820==    by 0x40EE40: server_thread (test_close_session.c:223)
==19820==    by 0x483F8B6: ??? (in /usr/lib64/valgrind/vgpreload_helgrind-amd64-linux.so)
==19820==    by 0x4888FAA: start_thread (in /lib64/libpthread-2.29.so)
==19820==    by 0x4D246AE: clone (in /lib64/libc-2.29.so)
==19820== 
==19820== This conflicts with a previous read of size 1 by thread #1
==19820== Locks held: none
==19820==    at 0x40EED7: np_start (test_close_session.c:239)
==19820==    by 0x48781E2: ??? (in /usr/lib64/libcmocka.so.0.5.1)
==19820==    by 0x4878A16: _cmocka_run_group_tests (in /usr/lib64/libcmocka.so.0.5.1)
==19820==    by 0x40F051: main (test_close_session.c:282)
==19820==  Address 0x42ce88 is 0 bytes inside data symbol "initialized"
==19820==
```

To my understanding all this says is that the `write` can happen completely independently of the `read` (before/during/after). As this is an atomic variable now, no partially-written/corrupted value can be read (so the `read` actually cannot happen during `write` even though it would still not matter, it is a flag so even partially-written value would be valid). So the problem reported should be that it is possible to `read` both `0` and `1`, it is impossible to know which. But that is exactly what is expected and the whole purpose of busy-waiting is to wait until the value changes without any expectation as to when it is going to happen. That is what I meant by "it should cause no harm".

Back to the general issue, I have gone only through *netopeer2* code and wanted to first fix it to my and your satisfaction before moving to other projects to save some repeated refactoring and fixing. Looking at all the warnings generated I could not find any regarding the original `initialized` variable problem, but maybe there never were any? Did you try to run the current tests to check whether they still occasionally hang? Like I said, that is what I would like to achieve first so that there is a reasonable chance that I understand the problem and know how to fix it.

As for all the warnings, I have not found any that would not make sense to me and most seem fairly easy to fix. However, as for the lock inversions (I know what that is) in *sysrepo*, these could potentially take days or even weeks to fix and since the current *sysrepo* is not a priority anymore (the new one should go into beta testing soon), I will not be fixing it now, possibly never. I would also prefer to keep the rest of those problems for later, for the same reason. What do you think, is it fine with you?

Regards,
Michal

**Comment by jktjkt** [comment]
> To my understanding all this says is that the `write` can happen completely independently of the
> `read` (before/during/after). As this is an atomic variable now, no partially-written/corrupted value 
> can be read (so the `read` actually cannot happen during `write` even though it would still not 
> matter, it is a flag so even partially-written value would be valid). So the problem reported should be 
> that it is possible to `read` both `0` and `1`, it is impossible to know which. But that is exactly what is 
> expected and the whole purpose of busy-waiting is to wait until the value changes without any 
> expectation as to when it is going to happen. That is what I meant by "it should cause no harm".

I do not understand that output, either. I think that memory operations around this `initialized` variable is now indeed correct. I checked valgrind's docs and it says that they support atomic ops. Perhaps this is something which is actually not supported? I have no idea.

Either way, I'm using a different tool and I no longer see this warning, which is a good thing. Thanks for fixing it!

> Back to the general issue, I have gone only through _netopeer2_ code and wanted to first fix it to my and your satisfaction before moving to other projects to save some repeated refactoring and fixing. Looking at all the warnings generated I could not find any regarding the original `initialized` variable problem, but maybe there never were any?

They've been fixed now; the logs I linked to today were from the latest devel branches of everything. Just FYI, here's an example on how these warnings used to look like prior to your fixes:

```
WARNING: ThreadSanitizer: data race (pid=10502)
  Write of size 4 at 0x000000f61038 by thread T1 (mutexes: read M21):
    #0 __wrap_nc_accept /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:180 (test_close_session+0x4f497a)
    #1 worker_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1400 (test_close_session+0x4f38b1)
    #2 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1758 (test_close_session+0x4f237a)
    #3 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f9971)

  Previous read of size 4 at 0x000000f61038 by main thread:
    #0 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:238 (test_close_session+0x4f4db8)
    #1 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #2 __libc_start_main ??:? (libc.so.6+0x24412)

  Location is global '<null>' at 0x000000000000 (test_close_session+0x000000f61038)

  Mutex M21 (0x000000f610f0) created at:
    #0 pthread_rwlock_init ??:? (test_close_session+0x442f77)
    #1 np2srv_init_schemas /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:964 (test_close_session+0x4f6022)
    #2 server_init /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1258 (test_close_session+0x4f2e38)
    #3 server_main /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/../main.c:1739 (test_close_session+0x4f2203)
    #4 server_thread /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:222 (test_close_session+0x4f9971)

  Thread T1 (tid=10504, running) created by main thread at:
    #0 pthread_create ??:? (test_close_session+0x42df5a)
    #1 np_start /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:236 (test_close_session+0x4f4d89)
    #2 _test_realloc ??:? (libcmocka.so.0+0x54e2)
    #3 __libc_start_main ??:? (libc.so.6+0x24412)

SUMMARY: ThreadSanitizer: data race /home/ci/src/cesnet-gerrit-public/CzechLight/dependencies/Netopeer2/server/tests/test_close_session.c:180 in __wrap_nc_accept
==================
```

> Did you try to run the current tests to check whether they still occasionally hang? Like I said, that is what I would like to achieve first so that there is a reasonable chance that I understand the problem and know how to fix it.

They are stable now. I suspect that the real issue was in fact that shared PID file. After I migrated to cmake-based timeouts and after adding atomics for `initialized`, but before I implemented separate PID files per each test, I saw one failure where the output said something like "server already running", and the test was stuck. Now I am not getting that even in repeated and parallel test runs.

> As for all the warnings, I have not found any that would not make sense to me and most seem fairly easy to fix. However, as for the lock inversions (I know what that is) in _sysrepo_, these could potentially take days or even weeks to fix and since the current _sysrepo_ is not a priority anymore (the new one should go into beta testing soon), I will not be fixing it now, possibly never.

Ack, this makes sense. Let's hope that they are rare enough in practice.

> I would also prefer to keep the rest of those problems for later, for the same reason. What do you think, is it fine with you?

Apart from the `lys_node_free` which looks dangerous to my untrained eyes, I'm OK with that. I'll therefore look into libnetconf2 and submit patches. I would still prefer if you could take a look at the `lys_node_free` race sooner because it is a new warning which I have not seen before -- and therefore I think it might have been introduced yesterday.

The original issue is now fixed, so I'm closing this one.

**Comment by michalvasko** [comment]
Regarding the `lys_node_free` warnings, all they mean is that due to some extreme reordering the *libyang* context could be freed before some data linked with it. I seriously doubt it can actually happen but I want to fix all these problems in *netopeer2*/*libnetconf2* eventually so I might as well look at this tomorrow. Will let you know.

Regards,
Michal
